### PR TITLE
6/create units

### PR DESCRIPTION
**Pull request by Hairash**
#6 

Add units

### CODE REVIEW

File: frontend/script.js [hunk 1]
```diff
@@ -16,14 +16,41 @@ function start() {
...
     grid.fill_hex(2, 5);
+
+    blue_unit = new Unit(app.loader.resources["blue unit"].texture, grid, 1, 3, {HP: 10, damage: 3});
+    red_unit = new Unit(app.loader.resources["red unit"].texture, grid, 2, 0, {HP: 5, damage: 4});
+    red_unit2 = new Unit(app.loader.resources["red unit"].texture, grid, 4, 5, {HP: 7, damage: 2});
+    // blue_unit2 = new Unit(app.loader.resources["blue unit"].texture, grid, 4, 5, {HP: 10, damage: 3});
+
+    blue_unit.move_to(4, 0);
+    red_unit2.move_to(3, 5);
+    // grid.hexs[0][0].addChild(blue_unit.sprite)
```
Comment (bronvic) [inline frontend/script.js:45]: something strange

File: frontend/script.js [hunk 2]
```diff
@@ -16,14 +16,41 @@ function start() {
...
+    // add the canvas that Pixi automatically created for you to the HTML document
     document.body.appendChild(app.view); // app.view - canvas element
 
+    // load images
+    app.loader.add('red unit', 'images/red unit.png');
+    app.loader.add('blue unit', 'images/blue unit.png');
+    app.loader.load(() => {loaded(app)});
+}
+
+function loaded(app) {
```
Comment (bronvic) [inline frontend/script.js:32]: maybe just load?

File: frontend/unit.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
+class Unit {
+    constructor(texture, grid, y, x, params) {
+        this.sprite = new PIXI.Sprite(texture);
+        this.grid = grid;
```
Comment (bronvic) [inline frontend/unit.js:4]: I will argue with this solution

It feels illogical that units can change environment (grid). When grid knows about unit and unit knows about grid bad things start happening. Program becomes overgrown with excessive dependencies. And data duplication (grid.hex[y][x] and unit.x, unit.y) leads to divergence of this data in future

That's why I suggest remove grid and coordinates from unit and left units referable from hexes (as you already did)

**Comment by bronvic** [comment]
And hey!

Where are the tests?)

File: frontend/script.js (hunk 2 above)
Comment (Hairash) [reply frontend/script.js:32]: loaded calls, when all images have been loaded

File: frontend/script.js (hunk 1 above)
Comment (Hairash) [reply frontend/script.js:45]: What?

File: frontend/unit.js (hunk 3 above)
Comment (Hairash) [reply frontend/unit.js:4]: ok, agree

**Comment by Hairash** [comment]
> And hey!
> 
> Where are the tests?)

Ok, will do

File: frontend/script.js (hunk 1 above)
Comment (alex-eg) [reply frontend/script.js:45]: Commented code, and without any comment on why it have been commented.
I'd say just looks a bit untidy.

File: frontend/unit.js (hunk 3 above)
Comment (alex-eg) [reply frontend/unit.js:4]: Second @bronvic

File: frontend/unit.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        this.set_new_coords(y, x);
+        this.hex.set_unit(this);
+    }
+
+    set_new_coords(y, x) {
+        this.y = y;
+        this.x = x;
+        this.hex = this.grid.hexs[y][x];
+    }
+}
```
Comment (alex-eg) [inline frontend/unit.js:39]: No newline at the end of the file.

Comment (bronvic) [reply frontend/unit.js:39]: shouldn't linter control this?

File: frontend/script.js (hunk 1 above)
Comment (Hairash) [reply frontend/script.js:45]: ok
removed

> Commented code, and without any comment on why it have been commented.
> I'd say just looks a bit untidy.
Ok
Removed

File: frontend/unit.js (hunk 4 above)
Comment (alex-eg) [reply frontend/unit.js]: It should, and it will.

File: frontend/script.js (hunk 2 above)
Comment (bronvic) [reply frontend/script.js]: makes sense now

File: frontend/unit.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    constructor(texture, img_size, params) {
+        this.sprite = new PIXI.Sprite(texture);
+        this.params = params;
+
+        this.sprite.anchor.set(0.5, 0.5);
+        this.scale_sprite(this.sprite, img_size);
+    }
+
+    // scale sprite to fit size
+    scale_sprite(sprite, size) {
```
Comment (bronvic) [inline frontend/unit.js:11]: you can access to sprite with this.sprite
no need to pass it as argument

File: frontend/game.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
+class Game {
+    // players_num - number of players (for the future)
+    // app - PIXI aplication, used for this game
+    // grid - grid for this game (Hex_grid)
+    constructor(players_num = 2, grid_params) {
```
[the remark this answers is not in the document]
Comment (Hairash) [reply frontend/game.js:5]: Yes. For the future

File: frontend/game.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        this.players_num = players_num;
+        this.app = null;
```
[the remark this answers is not in the document]
Comment (Hairash) [reply frontend/game.js:7]: ok

File: frontend/game.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        app.renderer.resize(window.innerWidth, window.innerHeight);
+
+        // add the canvas that Pixi automatically created for you to the HTML document
+        document.body.appendChild(app.view); // app.view - canvas element
+
+        this.app = app;
+    }
+
+    load_images() {
+        this.app.loader.add('red unit', 'images/red unit.png');
```
[the remark this answers is not in the document]
Comment (Hairash) [reply frontend/game.js:38]: your suggestions?

File: frontend/game.js [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    constructor(players_num = 2, grid_params) {
+        this.players_num = players_num;
+        this.app = null;
+        this.grid = null;
+    }
+
+    // needed, because constructor cannot be async
+    async start() {
+        this.create_stage();
+        await this.load_images();
```
[the remark this answers is not in the document]
Comment (Hairash) [reply frontend/game.js:14]: why?
It is necessary to start game.

File: frontend/game.js [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+            {HP: 10, damage: 3}, 1, 3);
+        let red_unit = this.create_unit(this.app.loader.resources["red unit"].texture, this.grid.hex_size,
+            {HP: 5, damage: 4}, 2, 0);
+        // let red_unit2 = new Unit(this.app.loader.resources["red unit"].texture, this.grid, 4, 5, {HP: 7, damage: 2});
+    }
+
+    // actions with units
+    create_unit(texture, img_size, params, y, x) {
+        let unit = new Unit(texture, img_size, params);
+        this.grid.hexs[y][x].set_unit(unit);
```
[the remark this answers is not in the document]
Comment (Hairash) [reply frontend/game.js:68]: No asserts yet.
We have an issue about it.

File: frontend/game.js [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+        return unit;
+    }
+
+    move_unit(from_y, from_x, to_y, to_x) {
+        let from_hex = this.grid.hexs[from_y][from_x];
+        let to_hex = this.grid.hexs[to_y][to_x];
+        let unit = from_hex.unit;
+
+        from_hex.unset_unit(unit);
```
[the remark this answers is not in the document]
Comment (Hairash) [reply frontend/game.js:77]: right

File: frontend/game.js [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    create_unit(texture, img_size, params, y, x) {
+        let unit = new Unit(texture, img_size, params);
+        this.grid.hexs[y][x].set_unit(unit);
+        return unit;
+    }
+
+    move_unit(from_y, from_x, to_y, to_x) {
+        let from_hex = this.grid.hexs[from_y][from_x];
+        let to_hex = this.grid.hexs[to_y][to_x];
+        let unit = from_hex.unit;
```
[the remark this answers is not in the document]
Comment (Hairash) [reply frontend/game.js:75]: No asserts yet.

File: frontend/unit.js (hunk 5 above)
Comment (Hairash) [reply frontend/unit.js:11]: ok

File: frontend/game.js [hunk 13]
```diff
@@ -0,0 +1,80 @@
+class Game {
```
[the remark this answers is not in the document]
Comment (Hairash) [reply frontend/game.js:1]: need to be discussed

File: frontend/game.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (bronvic) [reply frontend/game.js:14]: ok

File: frontend/game.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (bronvic) [reply frontend/game.js:5]: well ok

File: frontend/game.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (Hairash) [reply frontend/game.js:7]: linter don't allow this

**Pull request merged**